### ISSUE TITLE
Codes

### ISSUE THREAD

**Issue by archit6002-art** [archit6002-art/Archit-#2]
# CELL 1 — install (if on Colab)
!pip install -q scikit-learn imbalanced-learn joblib

# CELL 2 — imports
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, roc_auc_score, confusion_matrix
from imblearn.over_sampling import SMOTE
import joblib

# CELL 3 — load data (adjust path)
df = pd.read_csv('WA_Fn-UseC_-Telco-Customer-Churn.csv')  # from Kaggle

# quick clean
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')
df = df.dropna(subset=['TotalCharges'])
df['Churn'] = df['Churn'].map({'Yes':1,'No':0})

# features / target
X = df.drop(['customerID','Churn'], axis=1)
y = df['Churn']

# separate numeric / categorical
num_cols = X.select_dtypes(include=['int64','float64']).columns.tolist()
cat_cols = X.select_dtypes(include=['object']).columns.tolist()

# preprocessing
num_pipe = Pipeline([('scaler', StandardScaler())])
cat_pipe = Pipeline([('ohe', OneHotEncoder(handle_unknown='ignore', sparse=False))])

preproc = ColumnTransformer([('num', num_pipe, num_cols),
                             ('cat', cat_pipe, cat_cols)])

# create train/test split
X_train, X_test, y_train, y_test = train_test_split(X, y, 
                                                    test_size=0.2, 
                                                    stratify=y, 
                                                    random_state=42)

# apply preproc then SMOTE then model within pipeline-ish flow
X_train_prep = preproc.fit_transform(X_train)
X_test_prep  = preproc.transform(X_test)

# handle imbalance with SMOTE
sm = SMOTE(random_state=42)
X_res, y_res = sm.fit_resample(X_train_prep, y_train)

# train model
rf = RandomForestClassifier(n_estimators=200, random_state=42, n_jobs=-1)
rf.fit(X_res, y_res)

# eval
y_pred = rf.predict(X_test_prep)
y_proba = rf.predict_proba(X_test_prep)[:,1]

print(classification_report(y_test, y_pred))
print('ROC AUC:', roc_auc_score(y_test, y_proba))
print('Confusion matrix:\\n', confusion_matrix(y_test, y_pred))

# save model + preprocessor
joblib.dump({'preproc': preproc, 'model': rf}, 'churn_rf_pipeline.joblib')
